### PR TITLE
Added annotation to ignore None values in Case Classes

### PR DESCRIPTION
**Pull request by rozza**
SCALA-300

### CODE REVIEW

**Comment by rozza** [comment]
Needs documentation but wanted feedback on the approach first.

**Comment by JonathanMeyer2600** [comment]
I really do like the annotation approach. 
This is just what I thought about.

Maybe it should also toggle the "support missing optional values"?
So that without annotation you write and read null values and with annotation you do not?

**Comment by JonathanMeyer2600** [comment]
Okay, i just spoke to a colleague and he said: 
"having an annotation from mongo-scala-driver at our domain classes is not so good"

Maybe we can find a better place configure the "IgnoreNull"?

File: bson/src/test/scala/org/mongodb/scala/bson/codecs/MacrosSpec.scala [hunk 1]
```diff
@@ -141,6 +145,11 @@ class MacrosSpec extends FlatSpec with Matchers {
     decode(registry.get(classOf[OptionalValue]), buffer) should equal(OptionalValue("Bob", None))
   }
 
+  it should "be able to round trip ignored optional values" in {
+    roundTrip(OptionalValueIgnoreNone("Bob", None), """{name: "Bob"}""", classOf[OptionalValueIgnoreNone])
```
Comment (jyemin) [inline bson/src/test/scala/org/mongodb/scala/bson/codecs/MacrosSpec.scala]: Though Option<T> is preferred to null in Scala, the behavior of actual null values should be defined.  Currently both of these tests result in an NPE:

    roundTrip(OptionalValueIgnoreNone("Bob", null), """{"name": "Bob", value: null}""", classOf[OptionalValueIgnoreNone]);
    roundTrip(OptionalValueIgnoreNone(null, None), """{"name": null}""", classOf[OptionalValueIgnoreNone]);

File: bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala [hunk 2]
```diff
@@ -83,4 +84,23 @@ object Macros {
...
+     *
+     * {{{
+     * @IgnoreNone
+     * case class Person(name: String, nickName: Option[String])
+     *
+     * Person(name = "John", nickName: None) // becomes the equivalent of: { name: "John" }
+     * }}}
+     */
+    @meta.param
+    case class IgnoreNone() extends StaticAnnotation
```
Comment (jyemin) [inline bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala]: I'm able to add this annotation to an individual field, e.g.

    case class OptionalValueIgnoreNone(name: String,
                                       @IgnoreNone 
                                       value: Option[String])
 
As that currently has no effect, is there a way to make it a compiler error?

**Review by jyemin — CHANGES_REQUESTED** [review]

File: bson/src/test/scala/org/mongodb/scala/bson/codecs/MacrosSpec.scala (hunk 1 above)
Comment (rozza) [reply bson/src/test/scala/org/mongodb/scala/bson/codecs/MacrosSpec.scala]: `OptionalValueIgnoreNone("Bob", null)` should be left as is, it is pure scala and a user error to try and create that case class instance.

File: bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala (hunk 2 above)
Comment (rozza) [reply bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala]: Well that is a valid case class. 

We could make the macro fail in this instance but I'm not sure of the need to check annotations at the field level, with this commit.

File: bson/src/test/scala/org/mongodb/scala/bson/codecs/MacrosSpec.scala (hunk 1 above)
Comment (jyemin) [reply bson/src/test/scala/org/mongodb/scala/bson/codecs/MacrosSpec.scala]: What about the second one?

File: bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala (hunk 2 above)
Comment (jyemin) [reply bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala]: So nothing like java.lang.annotation.Target for Scala annotations?

Comment (rozza) [reply bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala]: Just checking now, I'm new to Scala annotations.

File: bson/src/test/scala/org/mongodb/scala/bson/codecs/MacrosSpec.scala (hunk 1 above)
Comment (rozza) [reply bson/src/test/scala/org/mongodb/scala/bson/codecs/MacrosSpec.scala]: Same its heresy in Scala, but added SCALA-301 to convert to a `CodecConfigurationException`

File: bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala (hunk 2 above)
Comment (rozza) [reply bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala]: No such thing in the Scala world.

File: bson/src/test/scala/org/mongodb/scala/bson/codecs/MacrosSpec.scala (hunk 1 above)
Comment (jyemin) [reply bson/src/test/scala/org/mongodb/scala/bson/codecs/MacrosSpec.scala]: Acknowledged

File: bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala (hunk 2 above)
Comment (jyemin) [reply bson/src/main/scala/org/mongodb/scala/bson/codecs/Macros.scala]: Acknowledged.

**Comment by rozza** [comment]
I rebased and this now sits on top of #38 

The changes can be seen in 063e73a. 

The upshot is you no longer need to annotate your case class to ignore None values.
Currently, I don't have any plans to add anymore annotations so we could remove it for now.

**Comment by rozza** [comment]
Ok final updates - removed the annotation, inclusion of None values is solely based on how the Codec is created.

**Review by jyemin — APPROVED** [review]

**Comment by rozza** [comment]
Rebased and merged

**Pull request closed without merging**